File: _submissions/19UCS159_Tharun/Day4/BSTiterator.java
public class BSTiterator {

  public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
      val = x;
    }
  }

  TreeNode root;

  public BSTiterator(TreeNode root) {
    this.root = helper(root);
  }

  private TreeNode helper(TreeNode root) {
    if (root == null)
      return null;
    TreeNode res = root;
    if (root.left != null) {
      TreeNode left = helper(root.left);
      res = left;
      while (left.right != null) {
        left = left.right;
      }
      root.left = null;
      left.right = root;
      res.right = helper(res.right);
      return res;
    } else {
      res.right = helper(root.right);
    }
    return res;
  }

  /** @return whether we have a next smallest number */
  public boolean hasNext() {
    return root != null;
  }

  /** @return the next smallest number */
  public int next() {
    int res = root.val;
    root = root.right;
    return res;
  }

  /**
   * Your Solution will be called like this: Solution i = new Solution(root);
   * while (i.hasNext()) System.out.print(i.next());
   */

}

File: _submissions/18UEI056_Prashun/Day13/BST_iterator.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
priority_queue<int,vector<int>,greater<int>>pq;
BSTIterator::BSTIterator(TreeNode *root) {
if(root==NULL) return;
queue<TreeNode*>q;
q.push(root);
while(!q.empty())
{ TreeNode*t=q.front();
q.pop();
pq.push(t->val);
if(t->left)
 q.push(t->left);
if(t->right)
q.push(t->right);
}
}

/** @return whether we have a next smallest number */
bool BSTIterator::hasNext() {
 if(pq.size()>0)
 return true;
 return false;
}

/** @return the next smallest number */
int BSTIterator::next() {
int x=0;
if(pq.size()>0)    
{x=pq.top();
pq.pop();}
return x;
}

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = BSTIterator(root);
 * while (i.hasNext()) cout << i.next();
 */
